### DISCUSSION TITLE
Override invidual linting rules when using a preset (recommended/all)

### DISCUSSION THREAD

**Discussion by awdyson** [biomejs/biome#838] [Ideas]
It'd be nice to be able to add exceptions to the linting presets.
Something like this:
```
  "linter": {
    "ignore": ["tsconfig.node.json"],
    "rules": {
      "all": true,
      "style": {
        "noImplicitBoolean": "off"
      }
    }
  }
```

Forgive me if this exists -- I couldn't figure out a great search phrase to find it 😅

**Comment by Conaclos** [comment]
This configuration should already provide what you want. Have you noticed an issue?

**Reply by awdyson** [reply]
Hmmm... must've dunno how I screwed that up, but you are correct.
Thanks for the reply.

**Reply by sassomedia** [reply]
@Conaclos Is this supposed to work for `"recommended": true` as well? 

In my case, I want to enable a few more rules that are not part of the recommended. As soon as I do that though, the recommended are no longer enabled. I'm using `v1.5.3` with the following config:

```json
{
  "$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
  "files": {
    "ignoreUnknown": true,
    "ignore": [".git/**", "dist/**", "locale/**", "node_modules/**", "public/**"],
    "include": [
      "config/**/*.js",
      "config/**/*.json",
      "cypress/**/*.js",
      "cypress/**/*.json",
      "scripts/**/*.js",
      "scripts/**/*.json",
      "src/**/*.js",
      "src/**/*.json"
    ]
  },
  "formatter": {
    "enabled": true,
    "lineWidth": 120
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "complexity": {
        "noExcessiveCognitiveComplexity": "error",
        "noVoid": "error",
        "useSimplifiedLogicExpression": "error"
      },
      "correctness": {
        "noUndeclaredVariables": "error",
        "noUnusedImports": "error"
      },
      "style": {
        "noNegationElse": "error",
        "useCollapsedElseIf": "error",
        "useFilenamingConvention": "error",
        "useNamingConvention": "error",
        "useShorthandAssign": "error"
      },
      "suspicious": {
        "noConsoleLog": "error",
        "noEmptyBlockStatements": "error",
        "useAwait": "error"
      }
    }
  },
  "organizeImports": { "enabled": true }
}
```

**Reply by Conaclos** [reply]
It should work. However, we had some bugs with presets that were foxed in the last patch releases. Could you check if you have the same issue with the last release 1.6.4?

**Reply by sassomedia** [reply]
Indeed, it does work with `1.6.4`! Unfortunately, the blocker preventing me from upgrading is https://github.com/biomejs/biome/issues/2085. That one is a pretty severe issue for any app containing JSX. Until that bug is fixed, I may have to go with OP's approach to turn `all` on and selectively turn off rules. That at least seems to work in `1.5.3`.

**Reply by Sec-ant** [reply]
> the blocker preventing me from upgrading is #2085. That one is a pretty severe issue for any app containing JSX.

@sassomedia We've seen many reports on #2085, so if it's a blocker for many users I'd like to fix it. Could you please share a minimal repro that can trigger the issue? Because I cannot reliably reproduce it and it may be editor-settings related.

**Reply by sassomedia** [reply]
@Sec-ant There is a basic [repro](https://biomejs.dev/playground/?code=ZgB1AG4AYwB0AGkAbwBuACAAZgAoACkAIAB7AAoAIAAgAHIAZQB0AHUAcgBuACAAPABkAGkAdgAvAD4AOwAKAH0A&typescript=false&jsx=false&script=true) already in the https://github.com/biomejs/biome/issues/2085 issue that displays the error. However, the OP mistakenly thinks there should be a better error message, when in fact, it should not display any error message at all since it is valid JSX. I see this error in my project as well, both in `.js` and `.jsx` files where the Biome parser displays a Typescript error despite my project not using Typescript.

It is possibly due to the reason @togami2864 mentioned [here](https://github.com/biomejs/biome/issues/2085#issuecomment-1997763516). As far as I can tell, JSX works fine with version `1.5.3`, any version after that exhibits the issue. Based on the change log, I don't see anything in `1.6.0` that could cause this. While there is mention of `useJsxKeyInIterable` being added as a rule, it is in "Nursery" which I have turned off. So that change seems unlikely to have caused issues.

Is there any other information that could help you troubleshoot?

**Reply by sassomedia** [reply]
As an aside; it would be great if the Playground would allow for selecting different versions of Biome, that would make it easier to see where things changed.

**Reply by Sec-ant** [reply]
@sassomedia The [playground link](https://biomejs.dev/playground/?code=ZgB1AG4AYwB0AGkAbwBuACAAZgAoACkAIAB7AAoAIAAgAHIAZQB0AHUAcgBuACAAPABkAGkAdgAvAD4AOwAKAH0A&typescript=false&jsx=false&script=true) you mentioned disables `TypeScript` and `JSX` features so it is expected that the file is parsed as a JavaScript only code and emits that error.

![image](https://github.com/biomejs/biome/assets/10386119/3fd60624-5109-4996-8a84-97015b0629f9)

A repro should be the error being emitted from a `.jsx` file.

**Reply by sassomedia** [reply]
@Sec-ant Good point, I missed that. I think you might be onto something with it being editor related. I just ran a file of mine with the Biome CLI:
```
% npx @biomejs/biome lint ./src/public.js
Checked 1 file in 7ms. No fixes needed.
```
and it works while VSCode still displays the error. It seems like this could be an issue specifically related to the VSCode plugin, since it points to `biome (parse)` rather than VSCode's Typescript parser:

<img width="860" alt="Screenshot 2024-04-08 at 20 37 56" src="https://github.com/biomejs/biome/assets/5614099/5b83e9b6-d528-4355-b6d7-fd1cd21ff6b8">

**Reply by Conaclos** [reply]
Dummy question: is it valid of using JSX inside a file ending with the `js` extension?

**Reply by sassomedia** [reply]
@Conaclos Yes, this is valid in a framework/library like React which internally converts JSX into a [Javascript equivalent](https://legacy.reactjs.org/docs/jsx-in-depth.html). Syntactically, it would be better to use a `.jsx` extension but a lot of React apps just don't.

In this case, even if I rename the file to `.jsx`, VSCode will still display the error. Only if I rename it to `.tsx`, a Typescript file with JSX, does the error go away. This could be a work-around but my project does not use Typescript at all, so it would be a bit difficult to justify renaming all files with Typescript extensions.

**Reply by Sec-ant** [reply]
> In this case, even if I rename the file to `.jsx`, VSCode will still display the error. 

This is the issue. I was hoping someone could give me a minimal repro. Because in my case the JSX syntax can be recognized correctly in a `.jsx` file. May I ask what does it display in the bottom right corner of VS Code when you open your `public.jsx` file?

![image](https://github.com/biomejs/biome/assets/10386119/a0b2ecfb-a3be-4b90-9273-c21a9859c264)

**Reply by sassomedia** [reply]
@Sec-ant Excellent question! I have the [Babel](https://marketplace.visualstudio.com/items?itemName=mgmcdermott.vscode-language-babel) plugin installed:

<img width="423" alt="Screenshot 2024-04-08 at 23 03 47" src="https://github.com/biomejs/biome/assets/5614099/60608682-fa7d-4ad3-bec2-20d77aec29b7">

If I disable Babel _**and**_ rename my file to `.jsx`, it does work! So there seems to be an incompatibility between those two plugins.

**Reply by Sec-ant** [reply]
Ah, this clears things out. The Biome language server uses language id to decide how to handle the file and that babel plugin will [associate `.jsx` files with the `javascript` language id](https://github.com/michaelgmcd/vscode-language-babel/blob/8b3a472748ad07c99dc022b66795c9eb46be4ccb/package.json#L63-L80) (and `javascriptreact` too, I don't know how the precedence works yet, maybe multiple language ids are reported and Biome only caches and uses one of them? idk yet), so the editor will report that file as a `javascript` file to the server and thus the error.

I think the expectation at least should be: the CLI and LSP should work in a consistent way. Though I kind of hesitate about supporting JSX syntax in a file with the language id `javascript` instead of `javascriptreact`. If that's a language id cache issue, that can be fixed easily. But if it is not, I wonder whether we have other ways to decide if there're JSX syntax in a file.

Have to do some debug now.

**Reply by Conaclos** [reply]
@Sec-ant  The language id is `babel JavaScript` here? Right? We could recognize this as JSX?

**Reply by Sec-ant** [reply]
> @Sec-ant The language id is `babel JavaScript` here? Right? We could recognize this as JSX?

That's the display name. The id is [`javascript`](https://github.com/michaelgmcd/vscode-language-babel/blob/8b3a472748ad07c99dc022b66795c9eb46be4ccb/package.json#L64).

**Reply by sassomedia** [reply]
@Sec-ant Ok, that makes sense! I wonder what changed between `1.5.3` and `1.6.0` that now interprets this differently. 🤔 Let me know if you need me to test anything further.

**Reply by Sec-ant** [reply]
I decided to fix it in an easy way: #2376

**Reply by sassomedia** [reply]
Wow, that was fast! Thank you for the quick fix!
